File: cpp/503.[M] Next Greater Element II.cpp
// https://leetcode.com/problems/two-sum/

#include <vector>
#include <unordered_map>
#include <assert.h>
#include <iostream>
#include <math.h>
#include <queue>

using namespace std;;

struct Comp {
    bool operator() (pair<int, int> a, pair<int, int> b) {
        return a.first > b.first;
    }
};

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int size = nums.size();
        if (size == 0)
            return {};
        priority_queue<pair<int, int>, vector<pair<int, int>>, Comp> pq = {};
        vector<int> res(size, -1);
        for (int i = 0; i < size; i++) {
            while (pq.size() > 0 && pq.top().first < nums[i]) {
                res[pq.top().second] = nums[i];
                pq.pop();
            }
            pq.push({nums[i], i});
        }
        for (int i = 0; i < size; i++) {
            while (pq.size() > 0 && pq.top().first < nums[i]) {
                res[pq.top().second] = nums[i];
                pq.pop();
            }
        }
        while (pq.size() > 0) {
            cout << pq.top().first << ' ';
            pq.pop();
        }
        cout << endl;
        return res;
    }
};

int main() {
    Solution sol = Solution();
    vector<int> input = {1,2,1};
    auto res = sol.nextGreaterElements(input);
    for (int n : res)
        cout << n << ' ';
    cout << endl;
}

File: python/503. [M] Next Greater Element II.py
# https://leetcode.com/problems/next-greater-element-ii/
from typing import List, Tuple


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        res = [0] * len(nums)
        stack = []  # type: List[Tuple[int, int]]
        for i, num in enumerate(nums):
            while stack and num > stack[-1][1]:
                res[stack[-1][0]] = num
                stack.pop()
            stack.append((i, num))

        for i, num in enumerate(nums):
            while stack and stack[-1][1] < num:
                res[stack[-1][0]] = num
                stack.pop()
            if not stack:
                break

        for i, num in stack:
            res[i] = -1

        return res

if __name__ == '__main__':
    sol = Solution()
    print(sol.nextGreaterElements([1,2,3,2,1]))
    # assert(sol.invertTree([1,2,1,2]) == 1)
